### ISSUE TITLE
Cache not clearing completely

### ISSUE THREAD

**Issue by patchimou** [luberda-molinet/FFImageLoading#624]
Hi,

I have a profile modal where i upload new pictures to my server. 
I upload it then i call `ImageService.InvalidateCacheEntryAsync()` to clear the cache, and reload it with `ReloadImage()`.

The problem is, the image reloads properly but when i dismiss the modal and open it again, the previous picture in cache shows up.

Does someone have a clue ?

**Comment by EmmanVazz** [comment]
You have CacheType set to All and removeSimilar set to true?

**Comment by patchimou** [comment]
Yes, I even tried to clear all the cache entirely, but the problem is the same

**Comment by daniel-luberda** [comment]
Hi @patchimou,

- Do you use newest prerelease?
- What are your image sources?
- Do you use custom cache keys somewhere? 
- Could you share some code?

**Comment by patchimou** [comment]
Hi @daniel-luberda,

I tried to reproduce it. I installed the newest prerelease v2.2.10-pre-403.
My image source is the url of my REST endpoint to get the image.
I don't use custom cache key anywhere, i have just a binding on the source (which is the url).

This is the main page i show the picture : 
```
<ffimageloading:CachedImage x:Name="ProfilePicture"
        WidthRequest="150" HeightRequest="150"
        DownsampleToViewSize="true"
        Source="{Binding ProfilePicture}">

        <ffimageloading:CachedImage.GestureRecognizers>
            <TapGestureRecognizer Command="{Binding OpenModal}"/>
        </ffimageloading:CachedImage.GestureRecognizers>
</ffimageloading:CachedImage>
```

When i click on it, it will open a modal with a simple layout where i see the picture and a button to take a picture with my phone : 
```
<StackLayout>
        <ffimageloading:CachedImage x:Name="ProfilePicture"
            WidthRequest="150" HeightRequest="150"
            Source="{Binding ProfilePicture}" />

        <Button Text="Upload new picture" 
                Command="{Binding TakePictureCommand}" />
</StackLayout>
```

So i take a picture, i upload it to my server. Then i call : 
`await ImageService.Instance.InvalidateCacheEntryAsync(ProfilePicture, CacheType.All, true); MessagingCenter.Send(this, "reloadPicture");`

which should invalidate the Cache of my ProfilePicture and all similar. Then i reload the image.

In conclusion, both image reload properly but if i dismiss the modal, go back to the main page et re open the modal, the previous image shows up, not the new one on my server.

**Comment by daniel-luberda** [comment]
Could you test the newest prerelease as there were some fixes with memory cache? (https://github.com/luberda-molinet/FFImageLoading/commit/9b88540b43083a19b2fa0f46011c02d68ca8fee2)

**Comment by patchimou** [comment]
I tried with the prerelease 405 and the problem is still the same.
I tried differently, instead of reading the image with the url, i used directly the stream of my image taken with the phone, thinking the problem was due to the url which is a string. But no luck either on this side.

I don't cache my pages because i always call a new page with my modal 
`OpenModal = new Command(() => navigation.PushModalAsync(new ModalUploadPicture())); `

**Comment by daniel-luberda** [comment]
@patchimou I understand now. It's probably not FFImageLoading issue, rather you http server caching your response, see this: https://github.com/luberda-molinet/FFImageLoading/issues/527#issuecomment-295366975

**Comment by patchimou** [comment]
I think I understood the problem. 
I manage to make it works by setting the image source to an empty string, clearing the cache and putting back the same url.

Then i thought that maybe the reloading effect or the clearing cache effect wouldn't takes effect if the old source was equals to the new one. Then i found in your code in   FFImageLoading.Forms.Droid\CachedImageRenderer.cs in the `UpdateBitmap (line 114)` method the following : 

> if (previous != null && _lastImageSource != null && ffSource != null && **!ffSource.Equals(_lastImageSource)**
> 				&& (string.IsNullOrWhiteSpace(placeholderSource?.Path) || placeholderSource?.Stream != null))
> 			{

**Comment by daniel-luberda** [comment]
Good catch :)  I'll fix it asap.

**Comment by daniel-luberda** [comment]
`ImageSourceBinding` has its own `Equals` override:

```C#
public override bool Equals(object obj)
{
    var item = obj as ImageSourceBinding;

    if (item == null)
    {
        return false;
    }

    return this.ImageSource == item.ImageSource && this.Path == item.Path && this.Stream == item.Stream;
}
``

As you see, there shouldn't be any problems for string based image sources (`ImageSource` is FFImageLoading `enum`) but I should change it to always return `false` for `Streams`

**Comment by patchimou** [comment]
All I know is that i need to reassign my image source with something different than the url. So i don't know if it's a normal behavior.

Anyway i managed to make it works that way so thank you for your help and your time, i appreciated :) !

**Comment by daniel-luberda** [comment]
`Stream` handling was fixed with https://github.com/luberda-molinet/FFImageLoading/commit/707030027df9354055d1e811c4ed9872b129eb8b

You can test it in the newest prerelease nuget.

**Comment by patchimou** [comment]
Tested it, I don't see any problems.
I will let you know in case of further problems
